d3d12/src/com.rs

```diff
@@ -0,0 +1,322 @@
+use crate::D3DResult;
+use std::{
+    fmt,
+    hash::{Hash, Hasher},
+    ops::Deref,
+    ptr,
+};
+use winapi::{ctypes::c_void, um::unknwnbase::IUnknown, Interface};
+
+/// A nullable pointer to a COM object.
+///
+/// # Invariants
+///
+/// This data structure contains one of the following:
+///
+/// 1. A null pointer.
+/// 2. A pointer to a valid instance of a COM object that implements `T`.
+#[repr(transparent)]
+pub struct ComPtr<T: Interface>(*mut T);
+
+impl<T: Interface> ComPtr<T> {
+    /// Creates a null ComPtr.
+    pub fn null() -> Self {
+        ComPtr(ptr::null_mut())
+    }
+
+    /// Constructs a tracked COM pointer from `raw`, calling [`AddRef`] on it.
+    ///
+    /// # Safety
+    ///
+    /// This constructor is unsound to use unless [invariants for this data structure][self] are
+    /// maintained by `raw`.
+    ///
+    /// [`AddRef`]: https://learn.microsoft.com/en-us/windows/win32/api/unknwn/nf-unknwn-iunknown-addref
+    pub unsafe fn from_raw(raw: *mut T) -> Self {
+        if !raw.is_null() {
+            (*(raw as *mut IUnknown)).AddRef();
+        }
+        ComPtr(raw)
+    }
+
+    /// Returns true if the inner pointer is null.
+    pub fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    /// Returns the raw inner pointer. May be null.
+    pub fn as_ptr(&self) -> *const T {
+        self.0
+    }
+
+    /// Returns the raw inner pointer as mutable. May be null.
+    pub fn as_mut_ptr(&self) -> *mut T {
+        self.0
+    }
+
+    /// Returns a pointer the inner pointer, which is casted as a pointer to [`c_void`].
+    ///
+    /// This is useful when D3D functions initialize objects by filling in a pointer to pointer
+    /// by taking `void**` as an argument.
+    ///
+    /// # Safety
+    ///
+    /// This method is not `unsafe` by itself. However, readers should remember that it is unsound
+    /// to assign to the pointee returned here unless [invariants for this data structure][self]
+    /// are maintained.
+    pub fn mut_void(&mut self) -> *mut *mut c_void {
+        let refer: *mut *mut T = &mut self.0;
+        let void: *mut *mut c_void = refer.cast();
+        void
+    }
```

Comment:
Why use cast? I think we shouldn't.
It is risky to change anything vs what we know worked before.